File: py_nodes/CommonNodeCode/metrics.py
import time

class MetricRecord:
    def __init__(self, seq_num, send_time):
        self.seq_num = seq_num
        self.send_time = send_time
        self.recv_time = 0
        self.latency = 0

class Metrics:
    MAX_RECORDS = 50

    def __init__(self):
        self.records = []

    def record_send(self, seq, timestamp=None):
        if len(self.records) >= self.MAX_RECORDS:
            self.records.pop(0)
        
        t = timestamp if timestamp is not None else int(time.time() * 1000)
        self.records.append(MetricRecord(seq,t))

    def record_receive(self, seq, timestamp=None):
        t = timestamp if timestamp is not None else int(time.time() * 1000)
        for record in self.records:
            if record.seq_num == seq:
                record.recv_time = t
                record.latency = record.recv_time - record.send_time
                break

    def calculate_average_latency(self):
        valid_records = [r.latency for r in self.records if r.recv_time != 0]
        if not valid_records:
            return 0
        return sum(valid_records) / len(valid_records)

File: IoTProjectLibrary/CommonNodeCode/metrics.cpp
#include "metrics.h"

void initMetrics(Metrics *m){
    m->count = 0;
}

void recordSend(Metrics *m, uint16_t seq, uint32_t time){
    if (m->count >= MAX_RECORDS) return;

    m->records[m->count].seq_num = seq;
    m->records[m->count].send_time = time;
    m->records[m->count].recv_time = 0;
    m->records[m->count].latency = 0;

    m->count++;
}

void recordReceive(Metrics *m, uint16_t seq, uint32_t time){
    for (int i = 0; i < m->count; i++){
        if (m->records[i].seq_num == seq){
            m->records[i].recv_time = time;
            m->records[i].latency = time - m->records[i].send_time;

            return;
        }
    }
}

uint16_t calculateAverageLatency(Metrics *m){
    if (m->count == 0) return 0;

    uint32_t total = 0;
    uint16_t valid = 0;

    for (int i = 0; i < m->count; i++){
        if (m->records[i].recv_time != 0){
            total += m->records[i].latency;
            valid++;
        }
    }

    if (valid == 0) return 0;
    return total/ valid;
}